Stride-Labs/stride
# Add Query for Pending ICQs

Closes: #XXX

## Context and purpose of the change
ICQs are not always picked up by the relayers and there's currently no transparency into their status - which can make debugging challenging. This PR adds a query that returns all pending ICQs to easily track down when a relayer has missed one.

## Brief Changelog
* Added a grpc query server to the `interchainquery` module
* Added a query for all pending ICQs
* Added query CLI command (`list-pending-queries`)
* Added query rest endpoint (`/Stride-Labs/stride/interchainquery/pending_queries`)

## Testing Plan
* Start the network with all host zones and start running the integration tests
```
make start-docker-all
make test-integration-docker
```
* Once the tests have started, run the following in a second terminal window. You should see pending queries appear.
```
while true; do echo "----------" && stridedl q interchainquery list-pending-queries && sleep 1; done
```
* Pause the gaia relayer (so the ICQs stop getting relayed)
```
docker pause dockernet-relayer-gaia-1
```
* Check the rest [URL](http://localhost:1317/Stride-Labs/stride/interchainquery/pending_queries) and verify that the gaia query is still there

## Author's Checklist

I have...

- [X] Run and PASSED locally all GAIA integration tests
- [ ] If the change is contentful, I either:
    - [ ] Added a new unit test OR 
    - [ ] Added test cases to existing unit tests
- [ ] OR this change is a trivial rework / code cleanup without any test coverage

If skipped any of the tests above, explain.


## Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the author wrote unit tests for new logic
- [ ] reviewed documentation exists and is accurate


## Documentation and Release Note

  - [ ] Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`?
  - [ ] This pull request updates existing proto field values (and require a backend and frontend migration)? 
  - [ ] Does this pull request change existing proto field names (and require a frontend migration)?
  How is the feature or change documented? 
      - [ ] jira ticket `XXX` 
      - [ ] specification (`x/<module>/spec/`) 
      - [ ] not documented
